numbat-wasm/www/index.js

```diff
@@ -110,11 +110,31 @@ function setup() {
   });
 }
 
+// TODO: I would prefer to use const lambdas rather than `function` in order to
+// avoid unexpected `this` resolution. Not sure if there's a reason this is
+// currently being used along with `var`. Those "old school" techniques do
+// hoist, but this can easily be worked around by reordering declarations.
+async function registerSw() {
+  if ("serviceWorker" in navigator) {
+    try {
+      // NOTE: The service worker URL should be stable between releases or this
+      // could lead to unexpected behavior.
+      await navigator.serviceWorker.register("./service-worker.js");
+    } catch (error) {
+      // NOTE: We don't allow the error to propagate because the app should
+      // still work without offline capabilities.
+      console.error("Failed to register Service Worker:", error);
+    }
+  } else {
+    console.warn("Service Workers not available in this browser");
+  }
+}
+
 var numbat;
 var combined_input = "";
 
 async function main() {
-  await init();
+  await Promise.all([init(), registerSw()]);
```

Comment:
I went back and forth as to whether this should be done in-line at `init()` time or during the `load` event [as is often suggested](https://web.dev/articles/service-workers-registration). My initial thought was to `register()` and `claim()` _first_ in order to take advantage of ServiceWorker-cached artifacts. However, it looks like the SW "prefetching" can actually [fall back to the browser cache](https://web.dev/articles/service-workers-registration#testing_things_out) if we allow the page to load first. This makes me think we should maybe wait until both the `load` event has fired _and_ the Numbat terminal has initialized before registering. On the other hand, this is a pretty minor detail given how large the WASM payload is at the moment and the fact that most users will be on modern devices with a good connection.

Reply:
I am mainly concerned about further delaying the time until a user can type something into the terminal. Due to the large WASM size, this is already a long wait time, unfortunately — you're right. If `registerSw` is negligible in comparison, I'm okay with either way.